presto-hive/src/test/java/io/prestosql/plugin/hive/util/TestThrottledAsyncQueue.java

```diff
@@ -0,0 +1,266 @@
+        assertTrue(queue.isFinished());
+    }
+
+    @Test(timeOut = 10_000)
+    public void testThrottleEmptyQueue()
+            throws Exception
+    {
+        // Make sure that the de-queuing is throttled even if we have enough elements in the queue
+
+        ThrottledAsyncQueue<Integer> queue = new ThrottledAsyncQueue<>(2, 10, executor);
+        assertTrue(queue.offer(1).isDone());
+        assertTrue(queue.offer(2).isDone());
+
+        // no throttling, enough elements in the queue
+        ListenableFuture<List<Integer>> future1 = queue.getBatchAsync(2);
+        assertTrue(future1.isDone());
+        assertEquals(getFutureValue(future1), ImmutableList.of(1, 2));
+        assertFalse(queue.isFinished());
+
+        // we are now throttled and the queue is empty
+        ListenableFuture<List<Integer>> future2 = queue.getBatchAsync(2);
+        assertFalse(future2.isDone());
+
+        Thread.sleep(1000L); // wait one second, after which we should not be throttled any more
+
+        // no batch is ready at that point as the queue is still empty
+        assertFalse(future2.isDone());
+
+        assertTrue(queue.offer(3).isDone());
+        assertTrue(queue.offer(4).isDone());
+        queue.finish();
+
+        assertEquals(getFutureValue(future2), ImmutableList.of(3, 4));
+
+        assertTrue(queue.isFinished());
+    }
+
+    @Test(timeOut = 10_000)
+    public void testBorrowThrows()
+            throws Exception
+    {
+        // It doesn't matter the exact behavior when the caller-supplied function to borrow fails.
+        // However, it must not block pending futures.
+
+        AsyncQueue<Integer> queue = new ThrottledAsyncQueue<>(100, 4, executor);
+        queue.offer(1);
+        queue.offer(2);
+        queue.offer(3);
+        queue.offer(4);
+        queue.offer(5);
+
+        ListenableFuture<?> future1 = queue.offer(6);
+        assertFalse(future1.isDone());
+
+        Runnable runnable = () -> {
+            getFutureValue(queue.borrowBatchAsync(1, elements -> {
+                throw new RuntimeException("test fail");
+            }));
+        };
+
+        try {
+            executor.submit(runnable).get();
+            fail("expected failure");
+        }
+        catch (ExecutionException e) {
+            assertContains(e.getMessage(), "test fail");
+        }
+
+        ListenableFuture<?> future2 = queue.offer(7);
+        assertFalse(future1.isDone());
+        assertFalse(future2.isDone());
+        queue.finish();
+        future1.get();
+        future2.get();
+        assertTrue(queue.offer(8).isDone());
+
+        try {
+            executor.submit(runnable).get();
+            fail("expected failure");
+        }
+        catch (ExecutionException e) {
+            assertContains(e.getMessage(), "test fail");
+        }
+
+        assertTrue(queue.offer(9).isDone());
+
+        assertFalse(queue.isFinished());
+        ArrayList<Integer> list = new ArrayList<>(queue.getBatchAsync(100).get());
```

Comment:
Why do we need to copy to an array list here?

Reply:
done (removed)